index.bs

```diff
@@ -4109,165 +4301,1226 @@ To <dfn>get the response data</dfn> given |response|:
+
+          1. Return [=error=] with [=error code=] [=invalid argument=].
+
+        1. If |codepoint| is U+005B ([), set |inside brackets| to true.
+
+        1. If |codepoint| is U+005D (]), set |inside brackets| to false.
+
+      1. Append |hostname| to |pattern url|.
+
+    1. Otherwise:
+
+      1. If |scheme| is not "<code>file</code>", append "<code>placeholder</code>" to
+         |pattern url|.
+
+      1. Set |has hostname| to false.
+
+    1. If |pattern| [=map/contains=] "<code>port</code>":
+
+      1. Let |port| be the result of [=trying=] to [=unescape
+         URL Pattern=] with |pattern|["<code>port</code>"].
+
+      1. Append "<code>:</code>" to |url pattern|.
+
+      1. For each |codepoint| in |port|:
+
+        1. If |codepoint| is not an [=ASCII digit=]:
+
+          1. Return [=error=] with [=error code=] [=invalid argument=].
+
+      1. Append |port| to |url pattern|.
+
+    1. If |pattern| [=map/contains=] "<code>pathname</code>":
+
+      1. Let |pathname| be the result of [=trying=] to [=unescape URL Pattern=] with
+         |pattern|["<code>pathname</code>"].
+
+      1. If |pathname| does not [=string/starts with|start with=] U+002F (/), then
+         append "<code>/</code>" to pattern.
+
+      1. For each |codepoint| in |pathname|:
+
+        1. If «U+003F (?), U+0023 (#)» [=set/contains=]
+           |codepoint|:
+
+          1. Return [=error=] with [=error code=] [=invalid argument=].
+
+      1. Append |pathname| to |url pattern|.
+
+    1. If |pattern| [=map/contains=] "<code>search</code>":
+
+      1. Let |search| be the result of [=trying=] to [=unescape URL pattern=] with
+         |pattern|["<code>search</code>"].
+
+      1. If |search| does not [=string/starts with|start with=] U+003F (?), then
+         append "<code>?</code>" to |url pattern|.
+
+      1. For each |codepoint| in |pathname|:
+
+        1. If |codepoint| is U+0023 (#):
+
+          1. Return [=error=] with [=error code=] [=invalid argument=].
+
+      1. Append |search| to |url pattern|.
+
+1. Otherwise, if |pattern| matches the <code>network.UrlPatternString</code> production:
+
+  1. Let |url pattern| be the result of [=trying=] to [=unescape URL
+     pattern=] with |pattern|["<code>pattern</code>"].
+
+1. Let |url| be the result of [=URL Parser|parsing=] |url pattern|.
+
+1. If |url| is failure, return [=error=] with [=error code=] [=invalid argument=].
+
+1. Let |parsed| be a struct with the following fields:
+   <dl>
+     <dt>protocol
+     <dd>|url|'s [=url/scheme=] if |has protocol| is true, or null otherwise.
+     <dt>hostname
+     <dd>|url|'s [=url/host=] if |has hostname| is true, or null otherwise.
+     <dt>port
+     <dd>[=Serialize an integer=] with |url|'s [=url/port=], if it's not null,
+         or null otherwise.
+     <dt>pathname
+     <dd>The result of running the [=URL path serializer=] with |url|'s
+         [=url/path=], if it's not [=list/empty=], or null otherwise.
```

Comment:
[`URL path serializer`](https://url.spec.whatwg.org/#url-path-serializer) takes a url, not a url path

Reply:
Apparently `path` can both be a string (opaque path) or a list (path), the current check only mentions list/empty. I think it can also be an empty string in the opaque path state, should we set null in that case as well?